clean-modular-code/write-pseudocode.md

```diff
@@ -0,0 +1,223 @@
+---
+```
+
+# Introduction to Pseudocode
+
+## Design an Efficient Data Workflow Using Pseudocode
+
+:::{admonition} What you'll learn
+:class: note
+
+* Be able to approach a coding task with a modular, systematic approach. 
+* Be able to write pseudocode.
+:::
+
+Pseudocode is an informal way of planning and structuring your code logic without worrying about the syntax of a specific programming language. It focuses on clearly expressing the steps a program will take to solve a problem. Writing pseudocode before diving into actual code helps clarify your thought process, spot logical gaps, and communicate your ideas to others.
+
+## Benefits of Pseudocode
+1. **Clarifies logic**: Helps you outline your program’s structure without getting bogged down by syntax.
+2. **Easier collaboration**: Allows you to communicate ideas with non-programmers or across different programming languages.
+3. **Quick debugging**: Spot logical errors early before writing actual code.
+
+## Using Pseudocode with LLMs
+
+LLMs (like ChatGPT) can assist in converting pseudocode into actual Python code. By writing clear pseudocode, you can prompt the LLM to generate clean, efficient Python code that follows your logic, helping you focus on problem-solving rather than syntax.
+
+## Example: Processing CrossRef Data from JOSS Papers
+
+You have inherited the code below, it's quite messy and hard to read 
+
+
+
+```{code-cell} ipython3
+from datetime import datetime
+
+# example list of papers with nested title, citations, and weird date format
+ps = [
+    {"title": ["P1"], "pub_date": "2023/05/10", "citations": [5]},
+    {"title": ["P2"], "pub_date": "2022/04/12", "citations": [3]},
+    {"title": ["P3"], "pub_date": "2021/03/15", "citations": [8]},
+]
+
+# process each paper manually 
+# P1
+p = ps[0]
+if p["citations"]:
+    pd1 = datetime.strptime(p["pub_date"], "%Y/%m/%d")
+    a = sum(p["citations"]) / len(p["citations"])
+    print({"tit": p["title"][0],"pd":pd1,"a":a})
+
+# P2
+x = ps[1]
+if x["citations"]:
+    pd2 = datetime.strptime(x["pub_date"], "%Y/%m/%d")
+    y = sum(x["citations"]) / len(x["citations"])
+    print({"t":x["title"][0],"pd":pd2,"y":y})
+
+# P3
+z = ps[2]
+if z["citations"]:
+    p3 = datetime.strptime(z["pub_date"],"%Y/%m/%d")
+    z_avg = sum(z["citations"]) / len(z["citations"])
+    print({"ttl":z["title"][0],"pdate":p3,"z_avg":z_avg})
+```
+
+
+
+### Step 1: Write Pseudocode
+Suppose you want to write a Python function that processes JSON data of JOSS papers from CrossRef, filtering out papers without valid DOIs and calculating citations per month.
+
+Here’s how you can start with pseudocode:
+
+* Load JSON data of JOSS papers from CrossRef.
+* Loop through each paper.
+    * ✅ Check if the DOI is valid.
+
+    * ✅ If valid, extract the number of citations and the publication date
+
+    * ✅ Calculate the number of citations per month since the publication date.
+   
+4.	Store results in a list.
+5.	Return the list of processed data.
```

Comment:
```suggestion
1. Load JSON data of JOSS papers from CrossRef.
2. Loop through each paper.
    * ✅ Check if the DOI is valid.

    * ✅ If valid, extract the number of citations and the publication date

    * ✅ Calculate the number of citations per month since the publication date.
   
3.	Store results in a list.
4.	Return the list of processed data.
```